tests/api/test_metrics_steps.py

```diff
@@ -0,0 +1,407 @@
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.NAME_KEY] == "step1"
+
+
+def test_metrics_steps_name_is_valid():
+    # given
+    metrics = CLIMetrics()
+    time_before_step = time.time()
+
+    # when
+    with metrics.track_step("step1"):
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.START_TIME_KEY] >= time_before_step
+    assert step1[_CLIMetricsStep.EXECUTION_TIME_KEY] > 0
+
+
+def test_metrics_steps_error_caught():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("step1"):
+        try:
+            raise RuntimeError()
+        except RuntimeError:
+            pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.ERROR_KEY] is None
+
+
+def test_metrics_steps_error_uncaught():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    try:
+        with metrics.track_step("step1"):
+            raise RuntimeError()
+    except RuntimeError:
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.ERROR_KEY] == "RuntimeError"
+
+
+def test_metrics_steps_no_parent():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("step1"):
+        pass
+
+    # then
+    assert len(metrics.steps) == 1
+    step1 = metrics.steps[0]
+    assert step1[_CLIMetricsStep.PARENT_KEY] is None
+
+
+def test_metrics_steps_with_parent_proper_names():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("parent"):
+        with metrics.track_step("child"):
+            pass
+
+    # then
+    assert len(metrics.steps) == 2
+    parent, child = metrics.steps
+
+    assert child[_CLIMetricsStep.NAME_KEY] == "child"
+    assert child[_CLIMetricsStep.PARENT_KEY] == "parent"
+    assert parent[_CLIMetricsStep.NAME_KEY] == "parent"
+
+
+def test_metrics_steps_with_parent_proper_ids():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("parent"):
+        with metrics.track_step("child"):
+            pass
+
+    # then
+    assert len(metrics.steps) == 2
+    parent, child = metrics.steps
+    assert child[_CLIMetricsStep.ID_KEY] != parent[_CLIMetricsStep.ID_KEY]
+    assert child[_CLIMetricsStep.PARENT_ID_KEY] == parent[_CLIMetricsStep.ID_KEY]
+    assert parent[_CLIMetricsStep.PARENT_ID_KEY] is None
+
+
+def test_metrics_steps_with_duplicate_names_proper_timing():
+    # given
+    metrics = CLIMetrics()
+
+    # when
+    with metrics.track_step("duplicate"):
+        with metrics.track_step("duplicate"):
+            pass
+
+    # then
+    assert len(metrics.steps) == 2
+    parent, child = metrics.steps
+
+    assert (
+        child[_CLIMetricsStep.START_TIME_KEY] > parent[_CLIMetricsStep.START_TIME_KEY]
```

Comment:
likely flaky, would use `>=` at least since there is no guarantee that execution will keep being slow enough to trigger a different start time for both steps.